core/app/models/spree/refund.rb

```diff
@@ -37,22 +38,57 @@ def description
       payment.payment_method.name
     end
 
-    private
-
-    # attempts to perform the refund.
+    # Must be called for the refund transaction to be processed.
+    #
+    # Attempts to perform the refund,
     # raises an error if the refund fails.
     def perform!
+      return true if perform_after_creation == false
       return true if transaction_id.present?
 
       credit_cents = money.cents
 
-      @response = process!(credit_cents)
+      response = process!(credit_cents)
+      log_entries.build(details: response.to_yaml)
 
-      self.transaction_id = @response.authorization
-      update_columns(transaction_id: transaction_id)
+      update!(transaction_id: response.authorization)
       update_order
     end
 
+    private
+
+    # This callback takes care of setting the behavior that determines if it is needed
+    # to execute the perform! callback after_create.
+    # Existing code that creates refund without explicitely passing
+    #
+    # perform_after_creation: false
+    #
+    # as attribute will still call perform! but a deprecation warning is emitted in order
+    # to ask users to change their code with the new supported behavior.
+    def set_perform_after_creation_default
+      return true if perform_after_creation == false
+
+      Spree::Deprecation.warn <<-WARN.strip_heredoc, caller
+        From Solidus v3.0 onwards, #perform! will need to be explicitly called when creating new
+        refunds. Please, change your code from:
+
+          Spree::Refund.create(your: attributes)
+
+        to:
+
+          Spree::Refund.create(your: attributes, perform_after_creation: false).perform!
```

Comment:
I have a small concern about this approach. In 3.0, `#perform!` will never be called after creation. That means the `perform_after_creation` attribute will become useless. Is our plan for 3.0 to introduce a deprecation warning when `perform_after_creation` is passed to `create`? We can't simply remove the attribute without a deprecation, because passing an unknown attribute to `create` will raise an error IIRC (and also because we want to make sure people clean up their code).

Reply:
That's the plan! In 3.0 we'll introduce a deprecation warning for that attribute so that people can remove that since it will be useless.

Reply:
Got it. I think that makes sense. Trying to figure out if there's a way to avoid introducing a new deprecation in 3.0 but I think this is the best way to do it.